discovery/tailscale/tailscale.go

```diff
@@ -0,0 +1,243 @@
+package tailscale
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"tailscale.com/client/tailscale"
+)
+
+const (
+	tsLabel                                 = model.MetaLabelPrefix + "tailscale_"
+	deviceLabel                             = tsLabel + "device_"
+	tsLabelDeviceAddresses                  = deviceLabel + "addresses"
+	tsLabelDeviceID                         = deviceLabel + "id"
+	tsLabelDeviceName                       = deviceLabel + "name"
+	tsLabelDeviceHostname                   = deviceLabel + "hostname"
+	tsLabelDeviceUser                       = deviceLabel + "user"
+	tsLabelDeviceClientVersion              = deviceLabel + "client_version"
+	tsLabelDeviceUpdate                     = deviceLabel + "update"
+	tsLabelDeviceOS                         = deviceLabel + "os"
+	tsLabelDeviceKeyExpiryDisavle           = deviceLabel + "key_expiry_disabled"
+	tsLabelDeviceAuthorized                 = deviceLabel + "authorized"
+	tsLabelDeviceIsExternal                 = deviceLabel + "is_external"
+	tsLabelDeviceMachineKey                 = deviceLabel + "machine_key"
+	tsLabelDeviceNodeKey                    = deviceLabel + "node_key"
+	tsLabelDeviceBlocksIncommingConnections = deviceLabel + "blocks_incomming_connections"
+	tsLabelDeviceEnabledRoutes              = deviceLabel + "enabled_routes"
+	tsLabelDeviceAdvertisedRoutes           = deviceLabel + "advertised_routes"
+	tsLabelDeviceTags                       = deviceLabel + "tags"
+	tsLabelDeviceTailnetLockKey             = deviceLabel + "tailnet_lock_key"
+)
+
+var (
+	DefaultSDConfig = SDConfig{
+		RefreshInterval:  model.Duration(60 * time.Second),
+		Port:             443,
+		URL:              "https://api.tailscale.com",
+		Tailnet:          "-",
```

Comment:
Why not an empty string?

Reply:
For tailnet selection, either a valid tailnet needs to be passed or `-` can be used to default to the "default" tailnet of the user. Leaving a blank here is possible, but would create a non-functional config.

source: https://github.com/tailscale/tailscale/blob/main/api.md#tailnet